src/holon/rule_engine/repositories/repository_base.py

```diff
@@ -18,13 +22,27 @@ def clone(self) -> RepositoryBaseClass:
 
         raise NotImplementedError()
 
-    # TODO ERIK
     def filter_model_subtype(self, model_subtype: Type) -> RepositoryBaseClass:
-        """
-        Keep only items in the repository that match with a certain filter
-        <RETURNS MODIFIED REPOSITORY>
-        """
-        raise NotImplementedError()
+        """Keep only items in the repository that are of the specified subtype, including further derived types."""
+
+        # Sanity check
+        self.__assert_valid_subtype(model_subtype)
+
+        objects = [o for o in self.objects if isinstance(o, model_subtype)]
+
+        clone = copy.copy(self)
+        clone.objects = objects
```

Comment:
Op branch #735 staat nu dat een repository wordt gecloned wanneer deze uit de ScenarioAggregate wordt gehaald (`get_repository_for_model_type()`), dus clonen in de methodes in een repository is niet meer nodig denk ik.
`self.objects` updaten en dan self returnen zou genoeg moeten zijn.

Reply:
Top! Één klein dingetje nog, het is nog goed als ie self returnt. Mijn idee was dat we dan filters etc. kunnen chainen. Maar dit kunnen we ook zo bespreken tijdens de progress meeting

Reply:
Als het doel van de methode is om het object te muteren vind ik niet dat hij zichzelf moet returnen. Wekt een verkeerde indruk.

Reply:
Dat is een goed punt, dat is niet zo netjes. Wel lijkt het me mooi als de functies van een `Repository` een bewerkt `Repository` object returnen, á la Pandas. 
Het gaat mij er meer om dat we niet elke stap hoeven te clonen, omdat dit tijd kost en de Repo in principe al gecloned wordt aan het begin van het filter proces. 
Wat vind je er van om in plaats van `self` te muteren, gewoon de constructor te gebruiken met een nieuwe set objects? Dus zo iets:
```
filtered_objects = [o for o in self.objects if isinstance(o, model_subtype)]
return self.__class__(objects=filtered_objects)
```

Op deze manier muteren we self niet, returnen we wel een geupdate repository en hoeven we niet te clonen

Reply:
Ik voorzie problemen met het behouden van invariants als het Repository zelf niet verantwoordelijk is om zich te clonen.

Maar kunnen we later nog aanpassen.

Reply:
Is w.m.b. niet blocking, niet zo relevant nu

Reply:
Sorry ik zag je bericht niet. Wat je zegt vindt ik wel elegant ja.

Reply:
Done.